dev/tools/localization/gen_l10n.dart

```diff
@@ -302,184 +307,220 @@ bool _isValidGetterAndMethodName(String name) {
+  final Set<LocaleInfo> supportedLocales = <LocaleInfo>{};
 
-          localeString = arbFilenameLocaleRE.firstMatch(entityPath)[1];
-        }
+  /// The class methods that will be generated in the localizations class
+  /// based on messages found in the template arb file.
+  final List<String> classMethods = <String>[];
 
-        arbFilenames.add(entityPath);
-        final LocaleInfo localeInfo = LocaleInfo.fromString(localeString);
-        if (supportedLocales.contains(localeInfo))
-          exitWithError(
-            'Multiple arb files with the same locale detected. \n'
-            'Ensure that there is exactly one arb file for each locale.'
-          );
-        supportedLocales.add(localeInfo);
-        supportedLanguageCodes.add('\'${localeInfo.languageCode}\'');
-      }
-    }
+  /// Initializes [l10nDirectory], [templateArbFile], [outputFile] and [className].
+  void initialize({
+    String l10nDirectoryPath,
+    String templateArbFileName,
+    String outputFileString,
+    String classNameString,
+  }) {
+    setL10nDirectory(l10nDirectoryPath);
+    setTemplateArbFile(templateArbFileName);
+    setOutputFile(outputFileString);
+    setClassName(classNameString);
   }
 
-  final List<String> classMethods = <String>[];
+  /// Sets the reference [Directory] for [l10nDirectory].
+  void setL10nDirectory(String arbPathString) {
+    if (arbPathString == null)
+      throw L10nException('Input string cannot be null');
+    l10nDirectory = _fs.directory(arbPathString);
+    if (!l10nDirectory.existsSync())
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, does not exist.\n"
+        'Make sure that the correct path was provided.'
+      );
+
+    final String l10nDirectoryStatModeString = l10nDirectory.statSync().modeString();
+    if (!_isDirectoryReadableAndWritable(l10nDirectoryStatModeString))
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, doesn't allow reading and writing.\n"
+        'Please ensure that the user has read and write permissions.'
+      );
+  }
 
-  Map<String, dynamic> bundle;
-  try {
-    bundle = json.decode(templateArbFile.readAsStringSync());
-  } on FileSystemException catch (e) {
-    exitWithError('Unable to read input arb file: $e');
-  } on FormatException catch (e) {
-    exitWithError('Unable to parse arb file: $e');
+  /// Sets the reference [File] for [templateArbFile].
+  void setTemplateArbFile(String templateArbFileName) {
+    if (templateArbFileName == null)
+      throw L10nException('Input string cannot be null');
+    if (l10nDirectory == null)
+      throw L10nException('l10nDirectory cannot be null when setting template arb file');
+
+    templateArbFile = _fs.file(path.join(l10nDirectory.path, templateArbFileName));
+    final String templateArbFileStatModeString = templateArbFile.statSync().modeString();
+    if (templateArbFileStatModeString[0] == '-' && templateArbFileStatModeString[3] == '-')
```

Comment:
IsDirectoryReadableAndWritable was factored out, maybe factor out `bool _isReadable(FileSystemEntity file)` too

Reply:
I cleaned up the function a little -- Decided to create `_isNotReadable` and `_isNotWritable` functions that are directly used by `setL10nDirectory` instead of nesting the logic.